addons/product/models/product_template.py

```diff
@@ -938,6 +939,76 @@ def _get_first_possible_combination(self, parent_combination=None, necessary_val
         """
         return next(self._get_possible_combinations(parent_combination, necessary_values), self.env['product.template.attribute.value'])
 
+    def _cartesian_product(self, product_template_attribute_values, parent_combination):
+        """
+        Generate all possible combination for attributes values (aka cartesian product).
+        It is equivalent to itertools.product except it skips invalid partial combinations before they are complete.
+
+        Imagine the cartesian product of 'A', 'CD' and range(1_000_000) and let's say that 'A' and 'C' are incompatible.
+        If you use itertools.product or any normal cartesian product, you'll need to filter out of the final result
+        the 1_000_000 combinations that start with 'A' and 'C' . Instead, This implementation will test if 'A' and 'C' are
+        compatible before even considering range(1_000_000), skip it and and continue with combinations that start
+        with 'A' and 'D'.
+
+        It's necessary for performance reason because filtering out invalid combinations from standard Cartesian product
+        can be extremely slow
+
+        :param product_template_attribute_values: the values we want all the possibles combinations of.
+        One list of values by attribute line
+        :return: a generator of product template values combinations (in the form of lists)
+        """
+        if not product_template_attribute_values:
+            raise StopIteration
+
+        all_exclusions = self._get_own_attribute_exclusions()
+        all_exclusions.update(self._get_parent_attribute_exclusions(parent_combination))
```

Comment:
I'm not sure to understand this line, why are those from the parent mixed up like this? They don't contain the same information.
Those from `_get_own_attribute_exclusions` have ptva of current product as key, but those from `_get_parent_attribute_exclusions` have ptva of parent product as key (which means they will never be "found" while iterating and therefore will likely not lead to any exclusion here).

Reply:
My bad. I expected the tests to test that, but they don't.

Reply:
Please add a test to cover every case that you noticed (will notice) of things that were not covered. This is the only way to safely apply suggestions (or make changes in the future) without breaking something that was fixed before.

Reply:
done